### ISSUE TITLE
Executing commnd in container hangs randomly (exec attach failed with error - connection reset by peer)

### ISSUE THREAD

**Issue by pgeorgiev333** [moby/moby#31421]
**Description**

I use Jenkins to orchestrate the process of starting a set of containers and run some functional tests against the product I'm testing using "docker-compose .... exec ....".  Sometimes the Jenkins job stops receiving the output from the test execution and just "hangs" as it still waits for the command to finish.

**Steps to reproduce the issue:**
1. Create docker-compose.yml with several containers and deploy them in the swarm cluster (i'm using the old swarm)
2. Run a command inside one of the containers:

**docker-compose -H <swarm_manager_ip>:3375 -p XXXXX exec -T executor script -qec 'command to run the tests' /dev/null**

I have a Jenkins pipeline that performs the above steps by executing them as shell commands for every commit so sometimes i have 3-4 jobs running in parallel and the issue happens only when i have multiple containers/jobs running in parallel.

**Describe the results you received:**

Sometimes one of the jobs seems to hang as the test output on the jenkins console stops. However when i login on the container i can see that the command has ended.

I 'm submitting this defect here because of the errors i see which come from dockerd:

**Feb 28 13:33:43 <swarm_node_1_IP> dockerd[12522]: time="2017-02-28T13:33:43.484639946+02:00" level=error msg="Error running exec in container: exec attach failed with error: write tcp <swarm_node_1_IP>:2375-><swarm_master_IP>:48280: write: connection reset by peer"
Feb 28 13:31:41 <swarm_node_1_IP> dockerd[12522]: time="2017-02-28T13:31:41.655800794+02:00" level=error msg="attach: stdout: write tcp <swarm_node_1_IP>:2375-><swarm_master_IP>:48280: write: connection reset by peer"**

**Describe the results you expected:**

I should see the output from the command execution in the jenkins console and the job that waits for the command to finish does not hang

**Additional information you deem important (e.g. issue happens only occasionally):**

I have one swarm master and two swarm agents. All docker-compose commands are executed against the swarm master on port 3375. The swarm agents are running on port 2375 on each machine.

**Output of `docker version`:**

( i have the same version on all 3 hosts)

```
Client:
 Version:      1.13.1
 API version:  1.26
 Go version:   go1.7.5
 Git commit:   092cba3
 Built:        Wed Feb  8 06:41:33 2017
 OS/Arch:      linux/amd64

Server:
 Version:      1.13.1
 API version:  1.26 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   092cba3
 Built:        Wed Feb  8 06:41:33 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 3
 Running: 3
 Paused: 0
 Stopped: 0
Images: 35
Server Version: 1.13.1
Storage Driver: devicemapper
 Pool Name: docker-thinpool
 Pool Blocksize: 524.3 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file:
 Metadata file:
 Data Space Used: 27.95 GB
 Data Space Total: 568.9 GB
 Data Space Available: 541 GB
 Metadata Space Used: 7.598 MB
 Metadata Space Total: 5.985 GB
 Metadata Space Available: 5.978 GB
 Thin Pool Minimum Free Space: 56.89 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: true
 Deferred Deletion Enabled: true
 Deferred Deleted Device Count: 0
 Library Version: 1.02.122 (2016-04-09)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: aa8187dbd3b7ad67d8e5e3a15115d3eef43a7ed1
runc version: 9df8b306d01f59d3a8029be411de015b7304dd8f
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.8.15-200.fc24.x86_64
Operating System: Fedora 24 (Twenty Four)
OSType: linux
Architecture: x86_64
CPUs: 24
Total Memory: 125.9 GiB
Name: <swarm_node_1_hostname>
ID: W5R5:WFIQ:NIUD:UM6Z:JICD:2Y3O:7LSI:D4YY:2OWO:6UMV:JDYF:636O
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
 com.axway.test.type=functional
Experimental: false
Cluster Store: etcd://<swarm_master_ip>:2379
Cluster Advertise: <swarm_node_1_IP>:2375
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

All these are physical machines.
The swarm version on all hosts  is 1.2.5
The docker-compose version on all hosts is:

docker-compose version 1.10.0, build 4bd6f1a
docker-py version: 2.0.1
CPython version: 2.7.9
OpenSSL version: OpenSSL 1.0.1t  3 May 2016

**Comment by thaJeztah** [comment]
Let me close this ticket for now, as it looks like it went stale.
